src/mds/nameserver2/writerlock.h

```diff
@@ -0,0 +1,121 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: 2022-07-25
+ * Author: YangFan (fansehep)
+ */
+
+#ifndef SRC_MDS_NAMESERVER2_WRITERLOCK_H_
+#define SRC_MDS_NAMESERVER2_WRITERLOCK_H_
+
+#include <string>
+#include <memory>
+
+#include "src/common/encode.h"
+#include "src/common/concurrent/name_lock.h"
+#include "src/common/lru_cache.h"
+#include "src/kvstorageclient/etcd_client.h"
+#include "src/mds/topology/topology_storge_etcd.h"
+#include "src/common/timeutility.h"
+
+namespace curve {
+namespace mds {
+
+
+using ::curve::common::NameLock;
+using ::curve::common::Uncopyable;
+using ::curve::kvstorage::EtcdClientImp;
+
+
+enum class Permission {
+    kReader = 0,
+    kWriter = 1,
+};
+
+enum class WriterStatus {
+    kSuccess = 0,
+    kError = 1,
+    kNoExist = 2,
+    kOutTime = 3,
+};
+
+
+class WriterLock {
+ public:
+    explicit WriterLock(const std::shared_ptr<EtcdClientImp>& etcdclientimp)
+    : etcdclient_(etcdclientimp) {}
+
+    /**
+     * client try to lock and get write permission
+     * if current file has a writer and no timeout,
+     * the client will be reader. Otherwise get
+     * write permission.
+     * @param:
+     * @return: return the client permission
+     */
+    Permission Lock(const std::string& filename,
+                    const std::string& clientuuid);
+
+    /**
+     * only the current file writer can unlock the file
+     * only old writer lost write permission, other client
+     * can get the current file write permission.
+     * @param:
+     * @return: unlock success return true
+     *          otherwise return false
+     */
+    bool Unlock(const std::string& filename,
+                const std::string& clientuuid);
+
+    /**
+     * only the current writer can update.
+     * writer mark the sesson with mds.
+     * @param:
+     * @return: update success return true, otherwise
+     *          return false
+     */
+    bool UpdateLockTime(const std::string& filename,
+                        const std::string& clientuuid);
+
+ private:
+    size_t Hash(const std::string& key);
+
+    WriterStatus GetWriterStatus(const std::string& filename,
+                                 std::string* clientuuid);
+
+    bool SetWriter(const std::string& filename,
+                   const std::string& clientuuid);
+
+    bool ClearWriter(const std::string& filename);
+
+    std::string EncodeKey(const std::string& writerinfo);
+
+    NameLock namelock_;
+    // writer can timeout 4250ms
+    static const uint64_t kWriterTimeOutopt_ = 4250;
```

Comment:
don't hard code it, a feasible way is to reuse `mds.file.expiredTimeUs` in mds.conf

Reply:
fix.